### PR TITLE
Issue 3223 - Ability to filter medatata

### LINKED ISSUE #3223: Limit metadata fields in Meta requests
### Description

There are 2 API requests which return long metadata lists: https://3drepo.github.io/3drepo.io/#api-Meta-queryMetadataByRules and https://3drepo.github.io/3drepo.io/#api-Meta-getAllMetadata

With some varied models & federations the number of Keys included across all elements can go into the thousands. This can make the request response take a long time and be very large (70MB seen on some models). Since a majority of the time the integrations are looking at specific Keys, we should give the users the option to add a filter to specify which keys should be returned.

### Goals

- [ ] As a user I want to query the latest model by default
- [ ] As a user I want to specify a revision of the model to query
- [ ] As a user I want to query Containers or Federations with the same endpoint
- [ ] As a user I want to include an array of Keys to filter by
- [ ] As a user I want to include hundreds of Keys in the filter array
- [ ] As a user some of my keys are long text fields
- [ ] As a user I still want to see the metadata _id and parents even if none of the keys matched so that I have a complete dataset
- 

### Related Resources

**Example**

Standard metadata response:
```
{
    "data": [
        {
            "_id": "005d7470-a66a-47a3-8c5a-cac53ac1b771",
            "parents": [
                "239cc6fb-4b07-4c22-90f5-d29b783ecb72"
            ],
            "metadata": {
                "IFC Type": "IfcBuildingElementProxy",
                "IFC GUID": "2wn9igZkDoJe6K_MZ5bAXW",
                "LayerName": "A-0233-M_Step33"
            }
        },
        {
            "_id": "0321aad0-1ca0-4ab2-88fb-be49b3400ce6",
            "parents": [
                "4817c549-c889-4929-8c08-eb748cc67e8d"
            ],
            "metadata": {
                "LayerName": "A-0120-M_Step20",
                "Floor": "Datum / Site Model",
                "RequiredForCOBieComponent": "True"
            }
        },
        {
            "_id": "0265184d-de0e-4405-8c19-c72fcbbc118b",
            "parents": [
                "8f20a60e-8f67-4421-93d9-b01a0accc36c"
            ],
            "metadata": {
                "Pen 1": "(do not use)",
                "Pen 2": "Black | 0.13",
                "Pen 3": "White Pen"
            }
        }
    ]
}
```

Query array:
`["IFC Type","LayerName"]`

Response:
```
{
    "data": [
        {
            "_id": "005d7470-a66a-47a3-8c5a-cac53ac1b771",
            "parents": [
                "239cc6fb-4b07-4c22-90f5-d29b783ecb72"
            ],
            "metadata": {
                "IFC Type": "IfcBuildingElementProxy",
                "LayerName": "A-0233-M_Step33"
            }
        },
        {
            "_id": "0321aad0-1ca0-4ab2-88fb-be49b3400ce6",
            "parents": [
                "4817c549-c889-4929-8c08-eb748cc67e8d"
            ],
            "metadata": {
                "LayerName": "A-0120-M_Step20"
            }
        },
        {
            "_id": "0265184d-de0e-4405-8c19-c72fcbbc118b",
            "parents": [
                "8f20a60e-8f67-4421-93d9-b01a0accc36c"
            ],
            "metadata": {
            }
        }
    ]
}
```

### PR DESCRIPTION
**Pull request by ChristosTsiotsias**
This fixes #3223

#### Description
Ability to filter returned metadata based on their keys has been added.

#### Test cases
Make a metadata request and pass the key filters in the query. Only the metadata objects that include those keys should be returned and only these keys and their values should be displayed.

- [ ] [Illegal characters in Query String](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134910277)
- [ ] [Filter not working on Metadata by Rules endpoint](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134905754)
- [ ] [Filter not applied to federations](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134904683)

### CODE REVIEW

File: backend/src/v4/models/scene.js [hunk 1]
```diff
@@ -75,12 +75,25 @@ Scene.findNodesByType = async function (account, model, branch, revision, type,
 	return findNodes(account, model, branch, revision, query, projection);
 };
 
+Scene.findMetadataNodesByFields = async function (account, model, branch, revision, fieldNames, projection = {}) {
+	const history = await History.getHistory(account, model, branch, revision);
+	const query = { $match: { rev_id: history._id, type: "meta" } };
+	projection = { $project: { ...projection, metadata: 1, parents: 1 } };
+
+	if (fieldNames.length) {
```
Comment (carmenfan) [inline backend/src/v4/models/scene.js:83]: I think it's better to assume the user may not pass in `[]` if they don't want to filter. 

```suggestion
	if (fieldNames?.length) {
```

Or you do `fieldNames=[]` on L78

**Review by carmenfan — CHANGES_REQUESTED** [review]

**Review by carmenfan — APPROVED** [review]

**Comment by carmenfan** [comment]
@mosment3DRepo  ready for QA

**Comment by mosment3DRepo** [comment]
## Filter not working on Metadata by Rules endpoint

As well as the Get All Metadata endpoint the filter was intended to work on the https://3drepo.github.io/3drepo.io/#api-Meta-queryMetadataByRules endpoint

**Review by mosment3DRepo — CHANGES_REQUESTED** [review]
see list

**Comment by carmenfan** [comment]
> ## Illegal characters in Query String
> The following model contains a metadata key which include a comma ("hello,world") when I run the following query I get no responses, unclear if there is any way to quote out the commas to avoid this happening
> 
> https://issue-3223.dev.3drepo.io/api/mosment3ddev/b60e7190-205d-11ec-a7d1-f183b9e6b73a/revision/master/head/meta/all.json?filter=hello,world

@mosment3DRepo  it still wouldn't work as Christos would have to do some more work to support it, but this should be passed in with the comma escaped, i.e.

> https://issue-3223.dev.3drepo.io/api/mosment3ddev/b60e7190-205d-11ec-a7d1-f183b9e6b73a/revision/master/head/meta/all.json?filter=hello%2Cworld

@ChristosTsiotsias  we'd have to run a pass to[ un-escape any characters](https://attacomsian.com/blog/nodejs-encode-decode-url) after splitting them into a string array

**Review by carmenfan — CHANGES_REQUESTED** [review]

**Comment by carmenfan** [comment]
Looks fine
@miadibe  @mosment3DRepo Christos has rewritten the query for metadata by rules slightly - we may want to run a few tests through that to make sure the results are still the same (with and without filter logic)

**Comment by miadibe** [comment]
- [ ] [Illegal Characters in query string](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134910277)
@mosment3DRepo to double check
- [ ] [Filter not working on Metadata by Rules endpoint](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134905754)
Tested and confirm it's working as expected 
- [ ] [Filter not applied to federations ](https://github.com/3drepo/3drepo.io/pull/3279#issuecomment-1134904683)
Tested and confirm is working as expected

**Review by mosment3DRepo — APPROVED** [review]
Tested issues raised in last review.

Tested all features from scratch and compared results to query on All Meta json

**Pull request merged**